### ISSUE TITLE
Add 'circ_seqs' argument to forgeBSgenomeDataPkgFromNCBI()

### ISSUE THREAD

**Issue by hpages** [Bioconductor/BSgenomeForge#13]
To do only once PR #12 is merged and issue #11 is closed. Make sure to resync your fork once PR #12 is merged.

**Background**

Many NCBI assemblies include sequences that are circular. For example, many assemblies include a mitochondrial sequence (usually called MT, but not always), and that sequence is normally circular. The BSgenome data packages produced by `forgeBSgenomeDataPkgFromNCBI()` should reflect that. Right now they don't. That's because we're setting the `CIRCSEQS` symbol to `"character(0)"`, so, as a consequence, no sequence is considered circular in the BSgenome data package produced by the function. We can see this by installing and loading the package, and by calling `seqinfo()` on the BSgenome object contained in it.  This returns a Seqinfo object where all the sequences have their circularity flags (`isCircular` field) set to `FALSE`.

We need to modify `forgeBSgenomeDataPkgFromNCBI()` so that it produces BSgenome data packages where the circularity flags are correct.

Note that, if we are forging a BSgenome data package for an assembly that is registered in the **GenomeInfoDb** package, we should be able to get the circularity information by just calling `getChromInfoFromNCBI(assembly_accession)`. However, in the case that the assembly is not registered, this won't help because this will return a data.frame where the `circular` column is filled with `NA`'s. So in this situation, we will need input from the user. We will get this input via a new argument, the `circ_seqs` argument.

**The `circ_seqs` argument**

This will be an optional argument (default value: `NULL`). Let's put it right before the `destdir` argument. It's value should be `NULL` or a character vector containing the names of the circular sequences.

More precisely:

- If `assembly_accession` is a registered assembly then we can infer `circ_seqs` from `getChromInfoFromNCBI(assembly_accession)`, so the user doesn't need to specify the argument. If they do, then we should check that what they supplied is in agreement with what we inferred, and raise an error if it's not.

- If `assembly_accession` is NOT a registered assembly, then the user must supply the names of the circular sequences via the `circ_seqs` argument. We should perform the following sanity checks on those names:
    - Make sure that they are actually valid sequence names for this assembly, and raise an error if they are not.
    - Make sure that they are the names of assembled molecules, and raise an error if they are not. The reason we do this is that only an assembled molecule can be circular.

As always, we should try to produce error messages that are as informative as possible.

**Comment by kakopo** [comment]
Hi Hervé. I have started at this issue but I am having a bit of a hard time trying to create conditionals based on whether the assembly is registered or not, such that the. I'd appreciate just a small pointer in this regard. Thank you!

**Comment by hpages** [comment]
You can use `registered_NCBI_assemblies()`. This returns a data.frame with 1 row per registered NCBI assembly:

    > library(GenomeInfoDb)
    > NCBI_assemblies <- registered_NCBI_assemblies()
    > head(NCBI_assemblies)
                 organism                  assembly       date        extra_info
    1 Acyrthosiphon pisum                  Acyr_1.0 2008/04/24                  
    2 Acyrthosiphon pisum                  Acyr_2.0 2010/06/24                  
    3 Acyrthosiphon pisum pea_aphid_22Mar2018_4r6ur 2019/05/20                  
    4 Ambystoma mexicanum           AmbMex13_14.1.0 2015/11/25 strain:laboratory
    5 Ambystoma mexicanum               ASM291563v1 2018/02/05      strain:DD151
    6 Ambystoma mexicanum               ASM291563v2 2018/12/04      strain:DD151
      assembly_accession circ_seqs
    1    GCA_000142985.1        MT
    2    GCA_000142985.2        MT
    3    GCA_005508785.1        MT
    4    GCA_001455525.1          
    5    GCA_002915635.1          
    6    GCA_002915635.2          

Then check whether the `assembly_accession` supplied by the user is in the `assembly_accession` column of this data frame. You can use the `%in%` operator for that e.g. `assembly_accession %in% NCBI_assemblies$assembly_accession` or `assembly_accession %in% NCBI_assemblies[ , "assembly_accession"]`.

**A note about the difference between `NCBI_assemblies$assembly_accession` and `NCBI_assemblies[ , "assembly_accession"]`:**

Both are valid ways to extract a column from a data frame. They are actually equivalent... well, almost! The only difference is how they handle an attempt at extracting a non-existing column. For example suppose that I make a typo in the column name e.g. I type `assembly_accessions` instead of `assembly_accession`. The first form won't say anything and will just return a `NULL`. The second form will fail with a clear error message.

Personally I prefer the 2nd form in my code. Here is why: I've seen situations where something like `x %in% df$foo` has been broken for years after a change somewhere in the upstream code resulted in the exact names of the columns in data frame `df` now being different. Problem is that `x %in% df$foo` kept working like nothing happened, except that now it would _always_ return `FALSE` (because `x %in% NULL` is always `FALSE`). This means that the function where `x %in% df$foo` is used was now silently returning wrong results. And it took years for someone to notice the problem :cry: 

That's why I prefer the 2nd form. If the exact names of the columns in `df` change, or if I mistype the name of the column, then `x %in% df[ , "foo"]` will fail. So I'll immediately know that my code is broken, and I'll immediately fix it :smiley: 

Note that the 1st form (`NCBI_assemblies$assembly_accession`) is still good when used _interactively_, because then you can use tab-completion, which is very convenient (especially for a long name like here) and also eliminates the risk of mistyping the column name.

**Comment by kakopo** [comment]
Hi @hpages.
Thank you for this earlier explanation. It really helped me get a starting point while writing the new helper function!
Unfortunately, I am still having a hard time getting the logic of this function through, especially in the instance that an assembly has no circular sequences. 
I receive this error when I run the code:
```
Error in if (all((unlist(circ_seqs_row) == FALSE))) character(0) : 
  missing value where TRUE/FALSE needed
```
which is very self explanatory but I don't know how to fix it yet. I'd appreciate any help, thank you!

**Comment by hpages** [comment]
> I receive this error when I run the code:
> ```
> Error in if (all((unlist(circ_seqs_row) == FALSE))) character(0) : 
>   missing value where TRUE/FALSE needed
> ```

You need to provide more context so we know what `circ_seqs_row` is and how it was obtained.

Copying the relevant code from [your last commit](https://github.com/Bioconductor/BSgenomeForge/pull/15/commits/15b98e72499b365295906a3771fe135db46f9027), this is how you obtain `circ_seqs_row`:

    seq_info <- getChromInfoFromNCBI(assembly_accession)
    seq_row <- seq_info[seq_info$circular == "TRUE", ]
    circ_seqs_row <- seq_row$SequenceName

So it looks like `circ_seqs_row` is a character vector containing _sequence names_. So now the question is: what is your `if` condition (`all((unlist(circ_seqs_row) == FALSE))`) testing for? What does it mean for _sequence names_ to be equal to FALSE? Also why unlist a character vector? (Have you tried that interactively? What does it do?)

Now for the particular error that you are getting, I guess that problems actually start earlier when you do `seq_info[seq_info$circular == "TRUE", ]`. Run this manually and observe what it does, especially in the case where `seq_info$circular` contains `NA`'s. Those `NA`'s are causing trouble when you do `seq_info[seq_info$circular == "TRUE", ]`, so then what happens later is not going to be good, especially when you do `if (all((unlist(circ_seqs_row) == FALSE)))` (look at what `all((unlist(circ_seqs_row) == FALSE))` returns, this is not something that `if` will be happy about, hence the error message you're getting).

Anyways, as we discussed earlier today during our call, extracting the circular sequences from the data frame returned by `getChromInfoFromNCBI()` is not something you need to do, nor want to do, if the assembly is not registered. Only when the assembly is registered. And in this case, you'll have the guarantee that `seq_info$circular` is `NA`-free, so you won't have to deal with those `NA`'s at all.

Finally, note that even if something like `if (all((unlist(x) == FALSE))) ...` seems to do the right thing, it's unnecessarily complicated/convoluted. The way we usually check that all the values in a _logicial vector_ `x` are FALSE (granted that this is what you are trying to do here), is with `if (!any(x)) ...` or `if (all(!x)) ...`, but the former is preferred.

Hope this helps.

Please let me know if I need to clarify things further or if you have any other doubt or question.

H.

**Comment by hpages** [comment]
I merged PR #15. However, after taking a closer look at `BSgenomeForge:::.get_circseqs()`, I realized that it's not exactly doing the right thing in the situation where the user specifies circular sequences that are a mix of assembled molecules with non-assembled molecules. For example, with assembly GCA_025782755.1:

    library(BSgenomeForge)

    seq_info <- getChromInfoFromNCBI("GCA_025782755.1")

    seq_info[1:4 , c("SequenceName", "SequenceRole")]
    #   SequenceName         SequenceRole
    # 1          III   assembled-molecule
    # 2           MT   assembled-molecule
    # 3   CP088301.1 unlocalized-scaffold
    # 4   CP088302.1 unlocalized-scaffold

    BSgenomeForge:::.get_circseqs("GCA_025782755.1", seq_info, circ_seqs=c("MT", "CP088301.1"))
    # [1] "MT"         "CP088301.1"

Remember that _all_ the circular sequences must be assembled molecules.

Of course we'll need to fix this. I still want to close this issue so maybe you could open a separate issue describing the problem? This way we won't forget (and we'll take care of it later).

Also please close this issue once you've created the issue for the above problem.

Thanks!

**Comment by kakopo** [comment]
Alright. I have created the issue.

**Comment by hpages** [comment]
@kakopo Thanks. Would be nice to include the link to the issue (you can you edit your comment above, no need to add a new comment for that). Thanks again!
